drivers/sensor/bme680/bme680.c

```diff
@@ -0,0 +1,444 @@
+	var1 = var1 >> 18;
+	var1 = ((32768 + var1) * (s32_t)data->par_p1) >> 15;
+	calc_press = 1048576 - adc_press;
+	calc_press = (s32_t)((calc_press - (var2 >> 12)) * ((u32_t)3125));
+	/* This max value is used to provide precedence to multiplication or
+	 * division in the pressure calculation equation to achieve least
+	 * loss of precision and avoiding overflows.
+	 * i.e Comparing value, signed int 32bit (1 << 30)
+	 */
+	if (calc_press >= (s32_t)0x40000000) {
+		calc_press = ((calc_press / var1) << 1);
+	} else {
+		calc_press = ((calc_press << 1) / var1);
+	}
+	var1 = ((s32_t)data->par_p9 *
+		(s32_t)(((calc_press >> 3)
+			 * (calc_press >> 3)) >> 13)) >> 12;
+	var2 = ((s32_t)(calc_press >> 2) * (s32_t)data->par_p8) >> 13;
+	var3 = ((s32_t)(calc_press >> 8) * (s32_t)(calc_press >> 8)
+		* (s32_t)(calc_press >> 8)
+		* (s32_t)data->par_p10) >> 17;
+
+	data->calc_press = (s32_t)(calc_press)
+			   + ((var1 + var2 + var3
+			       + ((s32_t)data->par_p7 << 7)) >> 4);
+}
+
+static void bme680_calc_humidity(struct bme680_data *data, u16_t adc_humidity)
+{
+	s32_t var1;
+	s32_t var2_1, var2_2, var2;
+	s32_t var3;
+	s32_t var4;
+	s32_t var5;
+	s32_t var6;
+	s32_t temp_scaled;
+	s32_t calc_hum;
+
+	temp_scaled = (((s32_t)data->t_fine * 5) + 128) >> 8;
+	var1 = (s32_t)(adc_humidity - ((s32_t)((s32_t)data->par_h1 * 16))) -
+	       (((temp_scaled * (s32_t)data->par_h3)
+		 / ((s32_t)100)) >> 1);
+	var2_1 = (s32_t)data->par_h2;
+	var2_2 = ((temp_scaled * (s32_t)data->par_h4) / (s32_t)100)
+		 + (((temp_scaled * ((temp_scaled * (s32_t)data->par_h5)
+				     / ((s32_t)100))) >> 6) / ((s32_t)100))
+		 +  (s32_t)(1 << 14);
+	var2 = (var2_1 * var2_2) >> 10;
+	var3 = var1 * var2;
+	var4 = (s32_t)data->par_h6 << 7;
+	var4 = ((var4) + ((temp_scaled * (s32_t)data->par_h7) /
+			  ((s32_t)100))) >> 4;
+	var5 = ((var3 >> 14) * (var3 >> 14)) >> 10;
+	var6 = (var4 * var5) >> 1;
+	calc_hum = (((var3 + var6) >> 10) * ((s32_t)1000)) >> 12;
+
+	if (calc_hum > 100000) { /* Cap at 100%rH */
+		calc_hum = 100000;
+	} else if (calc_hum < 0) {
+		calc_hum = 0;
+	}
+
+	data->calc_humidity = calc_hum;
+}
+
+static void bme680_calc_gas_resistance(struct bme680_data *data, u8_t gas_range,
+				       u16_t adc_gas_res)
+{
+	s64_t var1;
+	u64_t var2;
+	s64_t var3;
+
+	u32_t look_up1[16] = { 2147483647, 2147483647, 2147483647, 2147483647,
+			       2147483647, 2126008810, 2147483647, 2130303777,
+			       2147483647, 2147483647, 2143188679, 2136746228,
+			       2147483647, 2126008810, 2147483647, 2147483647 };
+
+	u32_t loop_up2[16] = { 4096000000, 2048000000, 1024000000, 512000000,
+			       255744255, 127110228, 64000000, 32258064,
+			       16016016, 8000000, 4000000, 2000000, 1000000,
+			       500000, 250000, 125000 };
+
+	var1 = (s64_t)((1340 + (5 * (s64_t)data->range_sw_err)) *
+		       ((s64_t)look_up1[gas_range])) >> 16;
+	var2 = (((s64_t)((s64_t)adc_gas_res << 15) - (s64_t)(16777216)) + var1);
+	var3 = (((s64_t)loop_up2[gas_range] * (s64_t)var1) >> 9);
```

Comment:
The slight differences in the gas resistance has adverse effects on the IAQ algorithms and hence it would be best to provide the data from the sensor in the best possible accuracy.